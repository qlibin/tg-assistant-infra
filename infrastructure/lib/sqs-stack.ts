import { Duration, Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { Alarm, TreatMissingData } from "aws-cdk-lib/aws-cloudwatch";
import { SnsAction } from "aws-cdk-lib/aws-cloudwatch-actions";
import {
  Effect,
  ManagedPolicy,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam";
import { Key } from "aws-cdk-lib/aws-kms";
import { Queue } from "aws-cdk-lib/aws-sqs";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { Topic } from "aws-cdk-lib/aws-sns";

export interface SQSStackProps extends StackProps {
  environment: string;
  projectName: string;
  enableEncryption?: boolean;
  enableCostOptimization?: boolean;
  watchTowerTopicArn?: string; // reserved for future cross-stack import
  orderQueueConfig?: {
    visibilityTimeoutSeconds?: number;
    maxReceiveCount?: number;
  };
  resultQueueConfig?: {
    visibilityTimeoutSeconds?: number;
    maxReceiveCount?: number;
  };
}

export class SQSStack extends Stack {
  public readonly orderQueue: Queue;
  public readonly orderDLQ: Queue;
  public readonly resultQueue: Queue;
  public readonly resultDLQ: Queue;

  public readonly queueEncryptionKey: Key;

  public readonly webhookRole: Role;
  public readonly workerRole: Role;
  public readonly feedbackRole: Role;

  public readonly queueAlertTopic: Topic;
  public readonly orderQueueAlarms: Alarm[] = [];
  public readonly resultQueueAlarms: Alarm[] = [];

  constructor(scope: Construct, id: string, props: SQSStackProps) {
    super(scope, id, props);

    const {
      environment,
      projectName,
      enableEncryption = true,
      orderQueueConfig,
      resultQueueConfig,
    } = props;

    // KMS key for SQS queues
    this.queueEncryptionKey = new Key(this, "QueueEncryptionKey", {
      description: `${projectName}-${environment} SQS encryption key`,
      enableKeyRotation: true,
    });

    // DLQs first
    this.orderDLQ = new Queue(this, "OrderDLQ", {
      queueName: `${projectName}-${environment}-order-dlq`,
      retentionPeriod: Duration.days(7),
      ...(enableEncryption
        ? { encryptionMasterKey: this.queueEncryptionKey }
        : {}),
    });

    this.resultDLQ = new Queue(this, "ResultDLQ", {
      queueName: `${projectName}-${environment}-result-dlq`,
      retentionPeriod: Duration.days(7),
      ...(enableEncryption
        ? { encryptionMasterKey: this.queueEncryptionKey }
        : {}),
    });

    // Main queues
    const orderVisibility = Duration.seconds(
      orderQueueConfig?.visibilityTimeoutSeconds ?? 300,
    );
    const orderMaxReceive = orderQueueConfig?.maxReceiveCount ?? 3;

    this.orderQueue = new Queue(this, "OrderQueue", {
      queueName: `${projectName}-${environment}-order`,
      visibilityTimeout: orderVisibility,
      retentionPeriod: Duration.days(14),
      receiveMessageWaitTime: Duration.seconds(6),
      deadLetterQueue: {
        queue: this.orderDLQ,
        maxReceiveCount: orderMaxReceive,
      },
      ...(enableEncryption
        ? { encryptionMasterKey: this.queueEncryptionKey }
        : {}),
    });

    const resultVisibility = Duration.seconds(
      resultQueueConfig?.visibilityTimeoutSeconds ?? 180,
    );
    const resultMaxReceive = resultQueueConfig?.maxReceiveCount ?? 3;

    this.resultQueue = new Queue(this, "ResultQueue", {
      queueName: `${projectName}-${environment}-result`,
      visibilityTimeout: resultVisibility,
      retentionPeriod: Duration.days(7),
      receiveMessageWaitTime: Duration.seconds(6),
      deadLetterQueue: {
        queue: this.resultDLQ,
        maxReceiveCount: resultMaxReceive,
      },
      ...(enableEncryption
        ? { encryptionMasterKey: this.queueEncryptionKey }
        : {}),
    });

    // IAM roles
    this.webhookRole = new Role(this, "WebhookLambdaRole", {
      roleName: `${projectName}-${environment}-webhook-role`,
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole",
        ),
      ],
      inlinePolicies: {
        OrderQueueProducer: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              resources: [this.orderQueue.queueArn],
              conditions: {
                StringEquals: { "aws:SourceAccount": this.account },
              },
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["kms:Decrypt", "kms:GenerateDataKey"],
              resources: [this.queueEncryptionKey.keyArn],
            }),
          ],
        }),
      },
    });

    this.workerRole = new Role(this, "WorkerLambdaRole", {
      roleName: `${projectName}-${environment}-worker-role`,
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole",
        ),
      ],
      inlinePolicies: {
        DualQueueWorkerAccess: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "sqs:ReceiveMessage",
                "sqs:DeleteMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueAttributes",
              ],
              resources: [this.orderQueue.queueArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["sqs:SendMessage", "sqs:GetQueueAttributes"],
              resources: [this.resultQueue.queueArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["kms:Decrypt", "kms:GenerateDataKey"],
              resources: [this.queueEncryptionKey.keyArn],
            }),
          ],
        }),
      },
    });

    this.feedbackRole = new Role(this, "FeedbackLambdaRole", {
      roleName: `${projectName}-${environment}-feedback-role`,
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole",
        ),
      ],
      inlinePolicies: {
        FeedbackDualQueueAccess: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "sqs:ReceiveMessage",
                "sqs:DeleteMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueAttributes",
              ],
              resources: [this.resultQueue.queueArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["sqs:SendMessage", "sqs:GetQueueAttributes"],
              resources: [this.orderQueue.queueArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["kms:Decrypt", "kms:GenerateDataKey"],
              resources: [this.queueEncryptionKey.keyArn],
            }),
          ],
        }),
      },
    });

    // Monitoring: alarms
    const stackName = `${projectName}-${environment}`;

    const orderQueueAgeAlarm = new Alarm(this, "OrderQueueAgeAlarm", {
      alarmName: `${stackName}-order-message-age`,
      alarmDescription: "Order messages aging in queue",
      metric: this.orderQueue.metricApproximateAgeOfOldestMessage({
        period: Duration.minutes(5),
      }),
      threshold: 900,
      evaluationPeriods: 2,
    });

    const resultQueueAgeAlarm = new Alarm(this, "ResultQueueAgeAlarm", {
      alarmName: `${stackName}-result-message-age`,
      alarmDescription: "Result messages aging in queue",
      metric: this.resultQueue.metricApproximateAgeOfOldestMessage({
        period: Duration.minutes(3),
      }),
      threshold: 600,
      evaluationPeriods: 2,
    });

    const orderDLQAlarm = new Alarm(this, "OrderDLQAlarm", {
      alarmName: `${stackName}-order-dlq-messages`,
      alarmDescription: "Failed orders in Dead Letter Queue",
      metric: this.orderDLQ.metricApproximateNumberOfMessagesVisible(),
      threshold: 1,
      evaluationPeriods: 1,
      treatMissingData: TreatMissingData.NOT_BREACHING,
    });

    const resultDLQAlarm = new Alarm(this, "ResultDLQAlarm", {
      alarmName: `${stackName}-result-dlq-messages`,
      alarmDescription: "Failed results in Dead Letter Queue",
      metric: this.resultDLQ.metricApproximateNumberOfMessagesVisible(),
      threshold: 1,
      evaluationPeriods: 1,
    });

    this.orderQueueAlarms.push(orderQueueAgeAlarm, orderDLQAlarm);
    this.resultQueueAlarms.push(resultQueueAgeAlarm, resultDLQAlarm);

    // SNS Topic and alarm actions
    this.queueAlertTopic = new Topic(this, "QueueAlertTopic", {
      topicName: `${stackName}-queue-alerts`,
      displayName: "SQS Alerts for Watch Tower",
      masterKey: this.queueEncryptionKey,
    });

    [
      orderQueueAgeAlarm,
      resultQueueAgeAlarm,
      orderDLQAlarm,
      resultDLQAlarm,
    ].forEach((alarm) => {
      alarm.addAlarmAction(new SnsAction(this.queueAlertTopic));
      alarm.addOkAction(new SnsAction(this.queueAlertTopic));
    });

    // SSM exports
    new StringParameter(this, "OrderQueueUrl", {
      parameterName: `/automation/${environment}/queues/order/url`,
      stringValue: this.orderQueue.queueUrl,
      description:
        "Order Queue URL for task distribution and requeue operations",
    });

    new StringParameter(this, "OrderQueueArn", {
      parameterName: `/automation/${environment}/queues/order/arn`,
      stringValue: this.orderQueue.queueArn,
      description: "Order Queue ARN for IAM permissions and monitoring",
    });

    new StringParameter(this, "ResultQueueUrl", {
      parameterName: `/automation/${environment}/queues/result/url`,
      stringValue: this.resultQueue.queueUrl,
      description: "Result Queue URL for processing results and feedback",
    });

    new StringParameter(this, "ResultQueueArn", {
      parameterName: `/automation/${environment}/queues/result/arn`,
      stringValue: this.resultQueue.queueArn,
      description: "Result Queue ARN for IAM permissions and monitoring",
    });

    new StringParameter(this, "QueueConfiguration", {
      parameterName: `/automation/${environment}/queues/config`,
      stringValue: JSON.stringify({
        orderQueue: {
          visibilityTimeout: orderVisibility.toSeconds(),
          maxReceiveCount: orderMaxReceive,
        },
        resultQueue: {
          visibilityTimeout: resultVisibility.toSeconds(),
          maxReceiveCount: resultMaxReceive,
        },
      }),
      description: "Queue configuration parameters for Lambda integration",
    });

    new StringParameter(this, "WebhookRoleArn", {
      parameterName: `/automation/${environment}/roles/webhook/arn`,
      stringValue: this.webhookRole.roleArn,
      description: "Webhook Lambda IAM Role ARN",
    });

    new StringParameter(this, "WorkerRoleArn", {
      parameterName: `/automation/${environment}/roles/worker/arn`,
      stringValue: this.workerRole.roleArn,
      description: "Worker Lambda IAM Role ARN",
    });

    new StringParameter(this, "FeedbackRoleArn", {
      parameterName: `/automation/${environment}/roles/feedback/arn`,
      stringValue: this.feedbackRole.roleArn,
      description: "Feedback Lambda IAM Role ARN",
    });

    new StringParameter(this, "QueueAlertTopicArn", {
      parameterName: `/automation/${environment}/monitoring/queue-alerts/topic-arn`,
      stringValue: this.queueAlertTopic.topicArn,
      description:
        "SNS Topic ARN for queue alerts integration with Watch Tower",
    });
  }
}
