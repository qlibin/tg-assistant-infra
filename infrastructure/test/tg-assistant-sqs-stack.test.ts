import * as cdk from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";
import { SQSStack } from "../lib/sqs-stack";

// AAA pattern tests for CDK stack

describe("SQS Stack", () => {
  const baseEnv = { account: "123456789012", region: "us-east-1" } as const;

  const makeStack = (
    overrides?: Partial<{
      envName: string;
      projectName: string;
      context?: Record<string, string>;
    }>,
  ) => {
    const appProps = overrides?.context
      ? { context: overrides.context as Record<string, unknown> }
      : undefined;
    const app = new cdk.App(appProps);

    return new SQSStack(app, "TestStack", {
      env: baseEnv,
      description: "Dual queue stack test",
      environment: overrides?.envName ?? "dev",
      projectName: overrides?.projectName ?? "tg-assistant",
    });
  };

  test("synthesizes expected CloudFormation template (snapshot)", () => {
    // Arrange
    const stack = makeStack({ envName: "dev" });

    // Act
    const fullTemplate = Template.fromStack(stack).toJSON() as Readonly<
      Record<string, unknown>
    >;

    // Assert - capture the entire synthesized template for full-stack coverage
    expect(fullTemplate).toMatchSnapshot();
  });

  test("creates four SQS queues (order, result, and their DLQs) with expected names", () => {
    // Arrange
    const stack = makeStack({ envName: "dev" });
    const template = Template.fromStack(stack);

    // Assert queue resources
    template.hasResourceProperties("AWS::SQS::Queue", {
      QueueName: "tg-assistant-dev-order-dlq",
    });
    template.hasResourceProperties("AWS::SQS::Queue", {
      QueueName: "tg-assistant-dev-result-dlq",
    });
    template.hasResourceProperties("AWS::SQS::Queue", {
      QueueName: "tg-assistant-dev-order",
    });
    template.hasResourceProperties("AWS::SQS::Queue", {
      QueueName: "tg-assistant-dev-result",
    });
  });

  test("enables KMS key rotation for SQS encryption key", () => {
    // Arrange
    const stack = makeStack({ envName: "dev" });
    const template = Template.fromStack(stack);

    template.hasResourceProperties("AWS::KMS::Key", {
      EnableKeyRotation: true,
    });
  });

  test("creates IAM roles with least-privilege inline policies for queues", () => {
    // Arrange
    const stack = makeStack({ envName: "dev" });
    const template = Template.fromStack(stack);

    // Role names
    template.hasResourceProperties("AWS::IAM::Role", {
      RoleName: "tg-assistant-dev-webhook-role",
    });
    template.hasResourceProperties("AWS::IAM::Role", {
      RoleName: "tg-assistant-dev-worker-role",
    });
    template.hasResourceProperties("AWS::IAM::Role", {
      RoleName: "tg-assistant-dev-feedback-role",
    });

    // At least one of the roles contains SQS/KMS actions
    template.hasResourceProperties(
      "AWS::IAM::Role",
      Match.objectLike({
        Policies: Match.anyValue(),
      }),
    );
  });

  test("exports queue URLs/ARNs, config, role ARNs, and SNS topic to SSM parameters", () => {
    // Arrange
    const stack = makeStack({ envName: "dev" });
    const template = Template.fromStack(stack);

    const paramNames = [
      "/automation/dev/queues/order/url",
      "/automation/dev/queues/order/arn",
      "/automation/dev/queues/result/url",
      "/automation/dev/queues/result/arn",
      "/automation/dev/queues/config",
      "/automation/dev/roles/webhook/arn",
      "/automation/dev/roles/worker/arn",
      "/automation/dev/roles/feedback/arn",
      "/automation/dev/monitoring/queue-alerts/topic-arn",
    ];

    for (const name of paramNames) {
      template.hasResourceProperties("AWS::SSM::Parameter", {
        Name: name,
      });
    }
  });

  test("creates SNS topic for alerts and wires CloudWatch alarms to it", () => {
    // Arrange
    const stack = makeStack({ envName: "dev" });
    const template = Template.fromStack(stack);

    template.hasResourceProperties("AWS::SNS::Topic", {
      TopicName: "tg-assistant-dev-queue-alerts",
    });

    // Four alarms present
    template.resourceCountIs("AWS::CloudWatch::Alarm", 4);
  });
});
