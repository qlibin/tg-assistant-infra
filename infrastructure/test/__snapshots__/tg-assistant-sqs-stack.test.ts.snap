// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQS Stack synthesizes expected CloudFormation template (snapshot) 1`] = `
{
  "Description": "Dual queue stack test",
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "FeedbackLambdaRole873CF38C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "sqs:ReceiveMessage",
                    "sqs:DeleteMessage",
                    "sqs:ChangeMessageVisibility",
                    "sqs:GetQueueAttributes",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "ResultQueue98CD34E0",
                      "Arn",
                    ],
                  },
                },
                {
                  "Action": [
                    "sqs:SendMessage",
                    "sqs:GetQueueAttributes",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "OrderQueue39B99167",
                      "Arn",
                    ],
                  },
                },
                {
                  "Action": [
                    "kms:Decrypt",
                    "kms:GenerateDataKey",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "QueueEncryptionKeyB11AE894",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "FeedbackDualQueueAccess",
          },
        ],
        "RoleName": "tg-assistant-dev-feedback-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "FeedbackRoleArn072A2029": {
      "Properties": {
        "Description": "Feedback Lambda IAM Role ARN",
        "Name": "/automation/dev/roles/feedback/arn",
        "Type": "String",
        "Value": {
          "Fn::GetAtt": [
            "FeedbackLambdaRole873CF38C",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "OrderDLQ87642B43": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "QueueEncryptionKeyB11AE894",
            "Arn",
          ],
        },
        "MessageRetentionPeriod": 604800,
        "QueueName": "tg-assistant-dev-order-dlq",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "OrderDLQAlarmC742FD20": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "QueueAlertTopic10D241EF",
          },
        ],
        "AlarmDescription": "Failed orders in Dead Letter Queue",
        "AlarmName": "tg-assistant-dev-order-dlq-messages",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "OrderDLQ87642B43",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "OKActions": [
          {
            "Ref": "QueueAlertTopic10D241EF",
          },
        ],
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "OrderQueue39B99167": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "QueueEncryptionKeyB11AE894",
            "Arn",
          ],
        },
        "MessageRetentionPeriod": 1209600,
        "QueueName": "tg-assistant-dev-order",
        "ReceiveMessageWaitTimeSeconds": 6,
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "OrderDLQ87642B43",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "OrderQueueAgeAlarmF610EF40": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "QueueAlertTopic10D241EF",
          },
        ],
        "AlarmDescription": "Order messages aging in queue",
        "AlarmName": "tg-assistant-dev-order-message-age",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "OrderQueue39B99167",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 2,
        "MetricName": "ApproximateAgeOfOldestMessage",
        "Namespace": "AWS/SQS",
        "OKActions": [
          {
            "Ref": "QueueAlertTopic10D241EF",
          },
        ],
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 900,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "OrderQueueArnCD29671B": {
      "Properties": {
        "Description": "Order Queue ARN for IAM permissions and monitoring",
        "Name": "/automation/dev/queues/order/arn",
        "Type": "String",
        "Value": {
          "Fn::GetAtt": [
            "OrderQueue39B99167",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "OrderQueueUrlD58C427B": {
      "Properties": {
        "Description": "Order Queue URL for task distribution and requeue operations",
        "Name": "/automation/dev/queues/order/url",
        "Type": "String",
        "Value": {
          "Ref": "OrderQueue39B99167",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "QueueAlertTopic10D241EF": {
      "Properties": {
        "DisplayName": "SQS Alerts for Watch Tower",
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "QueueEncryptionKeyB11AE894",
            "Arn",
          ],
        },
        "TopicName": "tg-assistant-dev-queue-alerts",
      },
      "Type": "AWS::SNS::Topic",
    },
    "QueueAlertTopicArnA604BDCE": {
      "Properties": {
        "Description": "SNS Topic ARN for queue alerts integration with Watch Tower",
        "Name": "/automation/dev/monitoring/queue-alerts/topic-arn",
        "Type": "String",
        "Value": {
          "Ref": "QueueAlertTopic10D241EF",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "QueueConfigurationE3E77BFC": {
      "Properties": {
        "Description": "Queue configuration parameters for Lambda integration",
        "Name": "/automation/dev/queues/config",
        "Type": "String",
        "Value": "{"orderQueue":{"visibilityTimeout":300,"maxReceiveCount":3},"resultQueue":{"visibilityTimeout":180,"maxReceiveCount":3}}",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "QueueEncryptionKeyB11AE894": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Description": "tg-assistant-dev SQS encryption key",
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "ResultDLQAB3B9549": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "QueueEncryptionKeyB11AE894",
            "Arn",
          ],
        },
        "MessageRetentionPeriod": 604800,
        "QueueName": "tg-assistant-dev-result-dlq",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "ResultDLQAlarmA62BED81": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "QueueAlertTopic10D241EF",
          },
        ],
        "AlarmDescription": "Failed results in Dead Letter Queue",
        "AlarmName": "tg-assistant-dev-result-dlq-messages",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "ResultDLQAB3B9549",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "OKActions": [
          {
            "Ref": "QueueAlertTopic10D241EF",
          },
        ],
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ResultQueue98CD34E0": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "QueueEncryptionKeyB11AE894",
            "Arn",
          ],
        },
        "MessageRetentionPeriod": 604800,
        "QueueName": "tg-assistant-dev-result",
        "ReceiveMessageWaitTimeSeconds": 6,
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "ResultDLQAB3B9549",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
        "VisibilityTimeout": 180,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "ResultQueueAgeAlarm240E7086": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "QueueAlertTopic10D241EF",
          },
        ],
        "AlarmDescription": "Result messages aging in queue",
        "AlarmName": "tg-assistant-dev-result-message-age",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "ResultQueue98CD34E0",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 2,
        "MetricName": "ApproximateAgeOfOldestMessage",
        "Namespace": "AWS/SQS",
        "OKActions": [
          {
            "Ref": "QueueAlertTopic10D241EF",
          },
        ],
        "Period": 180,
        "Statistic": "Maximum",
        "Threshold": 600,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ResultQueueArn37463251": {
      "Properties": {
        "Description": "Result Queue ARN for IAM permissions and monitoring",
        "Name": "/automation/dev/queues/result/arn",
        "Type": "String",
        "Value": {
          "Fn::GetAtt": [
            "ResultQueue98CD34E0",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ResultQueueUrl65EA052C": {
      "Properties": {
        "Description": "Result Queue URL for processing results and feedback",
        "Name": "/automation/dev/queues/result/url",
        "Type": "String",
        "Value": {
          "Ref": "ResultQueue98CD34E0",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "WebhookLambdaRole9BFB4813": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "sqs:SendMessage",
                    "sqs:GetQueueAttributes",
                    "sqs:GetQueueUrl",
                  ],
                  "Condition": {
                    "StringEquals": {
                      "aws:SourceAccount": "123456789012",
                    },
                  },
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "OrderQueue39B99167",
                      "Arn",
                    ],
                  },
                },
                {
                  "Action": [
                    "kms:Decrypt",
                    "kms:GenerateDataKey",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "QueueEncryptionKeyB11AE894",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "OrderQueueProducer",
          },
        ],
        "RoleName": "tg-assistant-dev-webhook-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "WebhookRoleArn672441D1": {
      "Properties": {
        "Description": "Webhook Lambda IAM Role ARN",
        "Name": "/automation/dev/roles/webhook/arn",
        "Type": "String",
        "Value": {
          "Fn::GetAtt": [
            "WebhookLambdaRole9BFB4813",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "WorkerLambdaRoleC56B01A8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "sqs:ReceiveMessage",
                    "sqs:DeleteMessage",
                    "sqs:ChangeMessageVisibility",
                    "sqs:GetQueueAttributes",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "OrderQueue39B99167",
                      "Arn",
                    ],
                  },
                },
                {
                  "Action": [
                    "sqs:SendMessage",
                    "sqs:GetQueueAttributes",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "ResultQueue98CD34E0",
                      "Arn",
                    ],
                  },
                },
                {
                  "Action": [
                    "kms:Decrypt",
                    "kms:GenerateDataKey",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "QueueEncryptionKeyB11AE894",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "DualQueueWorkerAccess",
          },
        ],
        "RoleName": "tg-assistant-dev-worker-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "WorkerRoleArn2B2D8264": {
      "Properties": {
        "Description": "Worker Lambda IAM Role ARN",
        "Name": "/automation/dev/roles/worker/arn",
        "Type": "String",
        "Value": {
          "Fn::GetAtt": [
            "WorkerLambdaRoleC56B01A8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
